service.go

```diff
@@ -0,0 +1,174 @@
+package ipam
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/microstorage"
+	"github.com/giantswarm/microstorage/memory"
+)
+
+const (
+	IPAMSubnetStorageKey       = "/ipam/subnet"
+	IPAMSubnetStorageKeyFormat = "/ipam/subnet/%s"
+)
+
+// Config represents the configuration used to create a new ipam service.
+type Config struct {
+	// Dependencies.
+	Logger  micrologger.Logger
+	Storage microstorage.Storage
+
+	// Settings.
+	// Network is the network in which all returned subnets should exist.
+	Network *net.IPNet
+}
+
+// DefaultConfig provides a default configuration to create a new ipam service
+// by best effort.
+func DefaultConfig() Config {
+	var err error
+
+	var newLogger micrologger.Logger
+	{
+		config := micrologger.DefaultConfig()
+		newLogger, err = micrologger.New(config)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	var newStorage microstorage.Storage
+	{
+		config := memory.DefaultConfig()
+		newStorage, err = memory.New(config)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	return Config{
+		// Dependencies.
+		Logger:  newLogger,
+		Storage: newStorage,
+	}
+}
+
+// New creates a new configured ipam service.
+func New(config Config) (*Service, error) {
+	// Dependencies.
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "logger must not be empty")
+	}
+	if config.Storage == nil {
+		return nil, microerror.Maskf(invalidConfigError, "storage must not be empty")
+	}
+
+	// Settings.
+	if config.Network == nil {
+		return nil, microerror.Maskf(invalidConfigError, "network must not be empty")
+	}
+
+	newService := &Service{
+		// Dependencies.
+		logger:  config.Logger,
+		storage: config.Storage,
+
+		// Settings.
+		network: *config.Network,
+	}
+
+	return newService, nil
+}
+
+type Service struct {
+	// Dependencies.
+	logger  micrologger.Logger
+	storage microstorage.Storage
+
+	// Settings.
+	network net.IPNet
+}
+
+// key returns a storage key for a given network.
+func key(network net.IPNet) string {
```

Comment:
Looks like marshal/unmarshal or encode/decode. unmarshal/decode is missing and I saw it inline in another method. I think it'd be good to extract and test if `o == decode(decode(o))`